File: examples/python/records.py
import mcapable

from common import write_sample_file


def main() -> None:
    path = write_sample_file()
    reader = mcapable.Reader.from_path(str(path))

    counts = {
        "Header": 0,
        "Footer": 0,
        "Schema": 0,
        "Channel": 0,
        "Message": 0,
        "Chunk": 0,
        "Other": 0,
    }

    stream = reader.records()
    for record in stream:
        kind = record.kind
        if kind in counts:
            counts[kind] += 1
        else:
            counts["Other"] += 1

    print("Record counts:")
    for key in ["Header", "Footer", "Schema", "Channel", "Message", "Chunk", "Other"]:
        print(f"  {key}:   {counts[key]}")


if __name__ == "__main__":
    main()

File: examples/typescript/records.ts
import { Reader } from "@mcapable/wasm";

import { sampleBytes } from "./common";

function main(): void {
  const reader = Reader.fromBytes(sampleBytes());
  const stream = reader.records();
  const counts = new Map<string, number>([
    ["Header", 0],
    ["Footer", 0],
    ["Schema", 0],
    ["Channel", 0],
    ["Message", 0],
    ["Chunk", 0],
    ["Other", 0],
  ]);

  for (;;) {
    const item = stream.next();
    if (!item) break;
    const record = item as { kind: string };
    const key = counts.has(record.kind) ? record.kind : "Other";
    counts.set(key, (counts.get(key) ?? 0) + 1);
  }

  console.log("Record counts:");
  for (const key of ["Header", "Footer", "Schema", "Channel", "Message", "Chunk", "Other"]) {
    console.log(`  ${key}: ${counts.get(key) ?? 0}`);
  }
}

main();
